### PR TITLE
[SEC-346] irep_ids: Adding in a NULL-object ID

### PR DESCRIPTION
**Pull request by JohnDumbell**
Originally I changed the existing null_object to an IREP_ID_TWO with the correct string in it, but then that would trip up anybody using "null_object" as a literal elsewhere, so for now reverted to having two identities instead.

Would you prefer me to change the existing null_object, leave it as-is in the current commit, or some other alternative?

### CODE REVIEW

**Review by tautschnig — CHANGES_REQUESTED** [review]
Thank you for raising this! I'm not sure what triggered your work on this - is there possibly a regression test that could be added? So why am I nevertheless requesting changes: I believe that all uses of "NULL-object" ought be replaced by the existing `ID_null_object`. Maybe @kroening can chime in as well?

**Comment by JohnDumbell** [comment]
It was just triggered by general cleanup - we were going to use the ID, noticed literals were being used and looked into adding it as an ID. I could write a quick test to enforce that the new (or old) irep_id == the old string literal though and maybe test one of the changed paths.

**Comment by tautschnig** [comment]
If you've got a use case for the ID then I think the absence of a new test is fine: the existing tests should guard against regressions anyway.

**Comment by smowton** [comment]
My only concern was perhaps some external tool parsing the string "NULL-object"; if you think that's unlikely shall we merge the two similar IDs?

**Comment by tautschnig** [comment]
If it's made an `IREP_ID_TWO` with "NULL-object" as the string representation then we should be fine (the only risk would be anyone using "null_object", but I don't think that's happening for anyone).

**Comment by JohnDumbell** [comment]
I've made the changes, but during a double-check I found another string which might be a target for changing, but I'm a lot less sure about this one. I've added the change in the second commit, if that should also be changed I'll merge it into the original commit, otherwise revert.

File: src/goto-programs/interpreter.cpp [hunk 1]
```diff
@@ -862,7 +862,7 @@ void interpretert::build_memory_map()
 {
   // put in a dummy for NULL
   memory.resize(1);
-  inverse_memory_map[0]="NULL-OBJECT";
+  inverse_memory_map[0]=ID_null_object;
```
Comment (tautschnig) [inline src/goto-programs/interpreter.cpp]: I think this change is fine. Strings used exactly once in the code base are very suspicious and should likely all be fixed.

File: src/analyses/ai.cpp [hunk 2]
```diff
@@ -506,7 +506,7 @@ bool ai_baset::do_function_call_rec(
     // We can't really do this here -- we rely on
     // these being removed by some previous analysis.
   }
-  else if(function.id()=="NULL-object")
+  else if(function.id()==ID_null_object)
```
Comment (tautschnig) [inline src/analyses/ai.cpp]: Nit pick (here and almost everywhere else in this PR): the clang-format Travis job has some suggestions to introduce whitespace around `==`. I'd follow that, but that's nit picking obviously.

**Review by tautschnig — APPROVED** [review]

**Comment by JohnDumbell** [comment]
Squashed commits and fixed the formatting.

**Review by owen-mc-diffblue — APPROVED** [review]

**Review by kroening — APPROVED** [review]

**Pull request merged**